src/mds/Locker.cc

```diff
@@ -594,6 +646,23 @@ bool Locker::acquire_locks(const MDRequestRef& mdr,
   return result;
 }
 
+/* Dropping *all* locks here is necessary so parent directory
+ * snap/layout/quiesce locks are unlocked for a future mksnap.  This is the
+ * primary purpose of the new quiescelock. An op, e.g. getattr, cannot block
+ * waiting for another lock held by quiesce_inode, e.g. filelock,
+ * which will prevent a mksnap on a subvolume inode (because getattr will
+ * already have gotten parent snaplocks, see Locker::try_rdlock_snap_layout).
+ */
+
+void Locker::handle_quiesce_failure(const MDRequestRef& mdr, std::string_view& marker)
```

Comment:
> Discussion: could we get into trouble here by not clearing the internal request state? Sometimes, operations encode their current progress in the `mdr` to consult it when retried. Dropping the locks here we hope that it will naturally make the request re-attempt to acquire them, but what if some internal flag was already set on the request, affecting the flow of the retry?

I think this question is all but mentioning the `mut->lock_state` but that is cleared in `drop_locks`. This is a common "start over" verse of code for an MDR so I don't really see the potential for a new issue arising from this.

> I wonder if there is some generic "clear" we could have called here to reset the request progress, and whether it would be safe.

I don't think it's a bad idea but any type of refactor should probably go ahead and split `MDRequestImpl` based on the client request / internal op so we can be more specific about the op state and a `clear` function that is understandable and well-defined.
 
> Rephrasing the concern: retriable flows may be unprepared to have their acquired locks dropped silently.

Reply:
OK, so we can leave this for the future.